File: Data Structures & Algorithms/regular-expression-matching/submission-4.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        n1, n2 = len(s), len(p)
        def dfs(i, j):
            if i >= n1 and j >= n2:
                return True

            if j >= n2:
                return False
            
            matching =  i < n1 and (s[i] == p[j] or p[j] == '.')
            if (j + 1 < n2 and p[j + 1] == '*'):
                return (dfs(i, j + 2) or
                       (matching and dfs(i + 1, j)))
            if matching:
                return dfs(i + 1, j + 1)
            
            return False
        
        return dfs(0, 0)


"""

s = aab
p = a*bc

"""

File: Data Structures & Algorithms/longest-increasing-path-in-matrix/submission-4.cs
public class Solution
{
    private int[][] directions = new int[][]
    {
        new int[] { 1, 0 },
        new int[] { -1, 0 },
        new int[] { 0, 1 },
        new int[] { 0, -1 }
    };
    public int LongestIncreasingPath(int[][] matrix)
    {
        int ROWS = matrix.Length, COLS = matrix[0].Length;
        int res = 0;
        for (int r = 0; r < ROWS; r++)
        {
            for (int c = 0; c < COLS; c++)
            {
                res = Math.Max(res, dfs(r, c, int.MinValue, matrix));
            }
        }
        return res;
    }

    public int dfs(int r, int c, int prevVal, int[][] matrix)
    {
        if (Math.Min(r, c) < 0
        || r >= matrix.Length
        || c >= matrix[0].Length
        || matrix[r][c] <= prevVal)
        {
            return 0;
        }

        int res = 1;
        foreach (var dir in directions)
        {
            res = Math.Max(res, 1 + dfs(r + dir[0],
                                    c + dir[1], matrix[r][c], matrix));
        }

        return res;
    }
}
